File: p1.py
# Array-2
# TC : O(N)
# SC : O(1)
## Problem1 (https://leetcode.com/problems/find-all-numbers-disappeared-in-an-array/)
class Solution:
    def findDisappearedNumbers(self, nums: List[int]) -> List[int]:
        # using no extra space:
        res = []
        for i in range(len(nums)):
            ind = abs(nums[i]) - 1
            if nums[ind]>0:
                nums[ind] = nums[ind] * -1
        for j in range(len(nums)):
            if nums[j] > 0:
                res.append(j+1)
        return res
        
        # # using O(1) space
        # k  = set(nums)
        # res = []
        # l = len(nums)
        # while l:
        #     if l not in k: # searching in set takes O(1), where as in list takes O(n)
        #         res.append(l)
        #     l -= 1
        # return res
    
            
        

File: p1.java
// Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

// Your code here along with comments explaining your approach

class Solution {
    public List<Integer> findDisappearedNumbers(int[] nums) {

        // Sorting the array in O(n)
        List<Integer> ans = sort(nums);
        return ans;
    }

    // Sort function
    public List<Integer> sort(int[] nums) {

        // Storing output in a list
        List<Integer> ans = new ArrayList<>();
        int len = nums.length;

        // Iterating through the array and swapping elements as we know the end position
        // of the elements
        for (int i = 0; i < len; i++) {
            while (nums[i] != nums[nums[i] - 1]) {
                swap(nums, i, nums[i] - 1);
            }
        }

        // Iterating through the elemtns and getting the position where the element does
        // not match with the position
        for (int i = 0; i < len; i++) {
            if (nums[i] != i + 1) {
                ans.add(i + 1);
            }
        }
        return ans;
    }

    // Swap function

    public void swap(int[] nums, int i, int j) {
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }
}
